File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l=0
        freq={}
        longest=0
        max_freq=0
        for r in range(len(s)):
            freq[s[r]]=freq.get(s[r],0)+1
            max_freq=max(max_freq,freq[s[r]])

            while (r-l+1) - max_freq > k:
                if s[l] in freq:
                    freq[s[l]]-=1
                    if freq[s[l]]<=0:
                        del freq[s[l]]
                l+=1
            longest = max(longest,r-l+1)
        return longest
            

        

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
         HashMap<Character, Integer> map = new HashMap<>();
         int st = 0;
         int max_len = 0;
         int curr = 0;
         int max_freq = 0;
         for(int end = 0; end<s.length();end++){
            char c = s.charAt(end);
            map.put(c,map.getOrDefault(c,0)+1);
            max_freq = Math.max(max_freq,map.get(c));
            while((end-st+1) - max_freq > k){
                map.put(s.charAt(st),map.get(s.charAt(st))-1);
                st++;
            }
            max_len = Math.max(max_len,end-st+1);
         }
    return max_len;
    }
}